tests/lib/rules/utils/char-source.js

```diff
@@ -0,0 +1,256 @@
+    () => {
+        const TESTS = [
+            {
+                description: "works with an empty string",
+                source: '""',
+                expectedCodeUnits: []
+            },
+            {
+                description: "works with surrogate pairs",
+                source: '"a𝄞z"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 2, source: "\ud834" },
+                    { start: 3, source: "\udd1e" },
+                    { start: 4, source: "z" }
+                ]
+            },
+            {
+                description: "works with escape sequences for single characters",
+                source: '"a\\x40\\u231Bz"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 2, source: "\\x40" },
+                    { start: 6, source: "\\u231B" },
+                    { start: 12, source: "z" }
+                ]
+            },
+            {
+                description: "works with escape sequences for code points",
+                source: '"a\\u{ffff}\\u{10000}\\u{10ffff}z"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 2, source: "\\u{ffff}" },
+                    { start: 10, source: "\\u{10000}" },
+                    { start: 10, source: "\\u{10000}" },
+                    { start: 19, source: "\\u{10ffff}" },
+                    { start: 19, source: "\\u{10ffff}" },
+                    { start: 29, source: "z" }
+                ]
+            },
+            {
+                description: "works with line continuations",
+                source: '"a\\\n\\\r\n\\\u2028\\\u2029z"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 11, source: "z" }
+                ]
+            },
+            {
+                description: "works with simple escape sequences",
+                source: '"\\"\\0\\b\\f\\n\\r\\t\\v"',
+                expectedCodeUnits: ['\\"', "\\0", "\\b", "\\f", "\\n", "\\r", "\\t", "\\v"]
+                    .map((source, index) => ({ source, start: 1 + index * 2 }))
+            },
+            {
+                description: "works with a <LS> character outside of a line continuation",
+                source: '"a\u2028z"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 2, source: "\u2028" },
+                    { start: 3, source: "z" }
+                ]
+            },
+            {
+                description: "works with a <PS> character outside of a line continuation",
+                source: '"a\u2029z"',
+                expectedCodeUnits: [
+                    { start: 1, source: "a" },
+                    { start: 2, source: "\u2029" },
+                    { start: 3, source: "z" }
+                ]
+            },
```

Comment:
These are valid since ES2019, see for example https://v8.dev/features/subsume-json.

```console
> console.log(eval('"a\u2028\u2029z"').length);

  4
```

Reply:
I misread the spec, thanks for the info!